### PR TITLE
Blocking/Unblocking Users + Refactoring

### PR DESCRIPTION
**Pull request by vianxnguyen**
## Overview

Implemented feature for blocking/unblocking users + refactoring. 

## Changes Made

- Implemented blocking/unblocking users feature (blocking from chatVC, profileVC, homeVC, unblocking from profileVC)
- Created module for presenting/removing pop up animations
- Combined MatchedUser & CommunityUser models
- Refactor FeedbackView into generalized OptionView

## Test Coverage

Tested on simulator. 

## Screenshots (delete if not applicable)
also blocking from homeVC & chatVC
https://user-images.githubusercontent.com/46759999/162554395-e4d1265f-1a9e-4ac1-93af-da9f1f8190b6.mp4

### CODE REVIEW

File: Pear/Models/UserV2.swift [hunk 1]
```diff
@@ -27,15 +27,14 @@ struct UserV2: Codable {
     let groups: [Group]
     let hasOnboarded: Bool?
     let pendingFeedback: Bool
-    let currentMatch: MatchV2?
+    var currentMatch: MatchV2?
```
Comment (mathewscullin) [inline Pear/Models/UserV2.swift:30]: Could this be changed back to `let` if the field isn't reassigned a value?

File: Pear/Models/UserV2.swift [hunk 2]
```diff
@@ -67,11 +66,13 @@ struct CommunityUser: Codable {
     let lastName: String
     let hometown: String
     let majors: [MajorV2]
-    let profilePicUrl: String?
+    var profilePicUrl: String?
```
Comment (mathewscullin) [inline Pear/Models/UserV2.swift:69]: Same here.

**Review by mathewscullin — APPROVED** [review]
Fantastic stuff! 🎉 Just some small nit callouts.

File: Pear/Assets.xcassets/optionicon.imageset/Contents.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    },
+    {
+      "filename" : "Bubble Drop Down (1).png",
+      "idiom" : "universal",
+      "scale" : "2x"
+    },
+    {
+      "filename" : "Bubble Drop Down (2).png",
+      "idiom" : "universal",
+      "scale" : "3x"
```
Comment (amyqxhuang) [inline Pear/Assets.xcassets/optionicon.imageset/Contents.json:16]: +1 to mat

File: Pear/Controller/ChatViewController.swift [hunk 4]
```diff
@@ -14,24 +14,27 @@ class ChatViewController: UIViewController {
...
     private let backButton = UIButton()
+    private let menuButton = UIButton()
     private let chatInputContainerView = UIView()
     private let chatInputTextField = UITextField()
     private let chatTableView = UITableView()
     private let emptyBackgroundView = UIView()
     private let emptyChatImage = UIImageView()
     private let sendMessageButton = UIButton()
     private let separatorView = UIView()
+    private var optionsMenuView: OptionsView?
```
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:25]: alphabetize

File: Pear/Controller/ChatViewController.swift [hunk 5]
```diff
@@ -14,24 +14,27 @@ class ChatViewController: UIViewController {
...
     // MARK: - Private Data Vars
     private let currentUser: UserV2
     private let databaseRef = Database.database().reference()
     private var dateKeys: [Date] = []
     private var groupedMessagesByDate: [[PearMessage]] = []
     private var messages: [PearMessage] = []
-    private let messageUser: MatchedUser
+    private let messageUser: CommunityUser
     private let status: String
+    private var displayMenu = true
```
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:35]: alphabetize

File: Pear/Controller/ChatViewController.swift [hunk 6]
```diff
@@ -62,6 +65,10 @@ class ChatViewController: UIViewController {
         backButton.setImage(UIImage(named: "backArrow"), for: .normal)
         backButton.addTarget(self, action: #selector(backPressed), for: .touchUpInside)
         navigationItem.leftBarButtonItem = UIBarButtonItem(customView: backButton)
+        
+        menuButton.setImage(UIImage(named: "optionicon"), for: .normal)
```
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:69]: camel case the name of option icon for consistency?

File: Pear/Controller/ChatViewController.swift (hunk 5 above)
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:35]: rename to `shouldDisplayMenu` for self-documentation

File: Pear/Controller/ChatViewController.swift [hunk 7]
```diff
@@ -281,6 +293,40 @@ class ChatViewController: UIViewController {
...
+                options: options,
+                superView: superView
+            )
+            
+            guard let optionsMenuView = optionsMenuView else { return }
+            view.addSubview(optionsMenuView)
+
+            optionsMenuView.snp.makeConstraints { make in
+                make.top.equalTo(menuButton.snp.bottom).offset(8)
+                make.trailing.equalTo(view.snp.trailing).offset(-25)
```
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:322]: `inset(25)` instead of offsetting by a negative number. Also, factor out the constants into variables. you can put them on line 305.

File: Pear/Controller/HomeViewController.swift [hunk 8]
```diff
@@ -168,15 +171,22 @@ class HomeViewController: UIViewController {
...
-            feedbackMenuView = FeedbackView(delegate: self)
+            guard let user = user, let superView = navigationController?.view else { return }
+            feedbackMenuView = OptionsView(
+                feedbackDelegate: self,
+                blockDelegate: self,
+                matchId: user.currentMatch?.id ?? 0,
+                blockId: user.currentMatch?.matchedUser.id ?? 0,
+                options: menuOptions,
+                superView: superView
+            )
```
Comment (amyqxhuang) [inline Pear/Controller/HomeViewController.swift:182]: very clean!

File: Pear/Controller/HomeViewController.swift [hunk 9]
```diff
@@ -168,15 +171,22 @@ class HomeViewController: UIViewController {
...
             guard let feedbackMenuView = feedbackMenuView else { return }
-            feedbackMenuView.layer.cornerRadius = 20
             view.addSubview(feedbackMenuView)
 
             feedbackMenuView.snp.makeConstraints { make in
                 make.top.equalTo(feedbackButton.snp.bottom).offset(5)
                 make.trailing.equalTo(view.snp.trailing).offset(-25)
-                make.size.equalTo(CGSize(width: 150, height: 130))
+                make.size.equalTo(CGSize(width: 150, height: 43 * menuOptions.count))
```
Comment (amyqxhuang) [inline Pear/Controller/HomeViewController.swift:189]: factor out constants into variables and do `.inset` instead of offsetting by a negative value.

File: Pear/Controller/HomeViewController.swift [hunk 10]
```diff
@@ -325,3 +340,17 @@ extension HomeViewController: ProfileMenuDelegate {
...
+        updateUserAndTabPage()
+    }
+    
+    func presentErrorAlert() {
+        present(UIAlertController.getStandardErrortAlert(), animated: true)
+    }
+    
+}
+
+
```
Comment (amyqxhuang) [inline Pear/Controller/HomeViewController.swift:356]: delete new line

File: Pear/Controller/MessagesViewController.swift [hunk 11]
```diff
@@ -18,7 +18,7 @@ class MessagesViewController: UIViewController {
 
     // MARK: - Private Data Vars
     private let databaseRef = Database.database().reference()
-    private var matchedUsers: [MatchedUser] = []
+    private var matchedUsers: [CommunityUser] = []
     private var matches: [TempMatchV2] = []
```
Comment (amyqxhuang) [inline Pear/Controller/MessagesViewController.swift:22]: talk to mat + noah about renaming all models with a V2? for example, `UserV2` should be `User` and `TempMatchV2` should be `TempMatch`

File: Pear/Controller/ChatViewController.swift (hunk 6 above)
Comment (amyqxhuang) [inline Pear/Controller/ChatViewController.swift:69]: talk to mat + noah about creating a file for static declarations of all the `UIImage`'s or something similar.

you can reference the way Volume did it [here](https://github.com/cuappdev/volume-ios/blob/master/Volume/Extensions/Image%2BExtensions.swift)

File: Pear/Controller/HomeViewController.swift [hunk 12]
```diff
@@ -30,6 +30,9 @@ class HomeViewController: UIViewController {
     private var displayMenu = true
     private let profileButtonSize = CGSize(width: 35, height: 35)
     private let tabs = ["Weekly Pear", "People"]
+    private var menuOptions = ["Send feedback", "Contact us", "Report user"]
```
Comment (amyqxhuang) [inline Pear/Controller/HomeViewController.swift:33]: similar to the comment about factoring out images, maybe do something similar with strings. I'm seeing that a lot of these strings are duplicated. 

you shouldn't worry about making these changes in this PR, bring it up to mat + noah and they can be a separate PR :)

File: Pear/Profile/ProfileViewController.swift [hunk 13]
```diff
@@ -62,37 +73,95 @@ class ProfileViewController: UIViewController {
...
+        unblockLabel.textColor = .darkGreen
+        view.addSubview(unblockLabel)
+        
+        let unblockLabelSize = CGSize(width: 280, height: 25)
+        let unblockButtonSize = CGSize(width: 196, height: 53)
+        
+        unblockLabel.snp.makeConstraints { make in
+            make.centerX.equalToSuperview()
+            make.centerY.equalToSuperview().offset(50)
+            make.size.equalTo(unblockLabelSize)
```
Comment (amyqxhuang) [inline Pear/Profile/ProfileViewController.swift:135]: why is this setting the size for labels? I feel like it could cut off the label for certain screen sizes? labels with only an x and y and will auto-size themselves to fit the text

File: Pear/Supporting/BlurEffectView.swift [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+import UIKit
+
+class BlurEffectView: UIVisualEffectView {
+    
+    private let animator = UIViewPropertyAnimator(duration: 1, curve: .linear)
+    
+    override func didMoveToSuperview() {
+        guard let superview = superview else { return }
+        backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.3)
```
Comment (amyqxhuang) [inline Pear/Supporting/BlurEffectView.swift:17]: I think you can do `UIColor.black.withAlphaComponent(0.3)`

**Review by amyqxhuang — CHANGES_REQUESTED** [review]
love the work so far! added some feedback :)

**Review by amyqxhuang — APPROVED** [review]
looks good! 🍐🍐
